pkg/util/httpgrpcutil/middleware/grpc_instrumentation.go

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	"github.com/uber/jaeger-client-go"
+)
+
+const httpGRPCHandleMethod = "/httpgrpc.HTTP/Handle"
+const httpSpanNameSep = " "
+
+// InitHTTPMiddleware initializes gorilla/mux-compatible HTTP middleware
+func InitHTTPMiddleware(router *mux.Router) {
+	middleware := OpenTracingHTTPGRPCMiddleware{router: router}
+	router.Use(middleware.Wrap)
+}
+
+type OpenTracingHTTPGRPCMiddleware struct {
+	router *mux.Router
+}
+
+// Wrap creates and decorates server-side tracing spans for httpgrpc requests
+//
+// The httpgrpc client wraps HTTP requests up into a generic httpgrpc.HTTP/Handle gRPC method.
+// The httpgrpc server unwraps httpgrpc.HTTP/Handle gRPC requests into HTTP requests
+// and forwards them to its own internal HTTP router.
+//
+// By default, the server-side tracing spans for the httpgrpc.HTTP/Handle gRPC method
+// have no data about the wrapped HTTP request being handled.
+//
+// Wrap starts & decorates a child span with OTEL HTTP server span name and tag conventions
+// and attaches the HTTP server span tags to the parent httpgrpc.HTTP/Handle gRPC span,
+// allowing tracing tooling to differentiate the HTTP requests represented by the httpgrpc.HTTP/Handle spans.
+//
+// Parent span tagging depends on using a Jaeger tracer for now in order to access OperationName(),
+// which is not required to satisfy the generic opentracing Tracer interface.
+//
+// (Currently-experimental) OTEL HTTP tracing standards are used, as opentracing has limited standards defined.
+func (m *OpenTracingHTTPGRPCMiddleware) Wrap(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		tracer := opentracing.GlobalTracer()
+
+		matchedRoute, _ := mux.CurrentRoute(r).GetPathTemplate()
+
+		parentSpan := opentracing.SpanFromContext(ctx)
+		if parentSpan, ok := parentSpan.(*jaeger.Span); ok {
+			if parentSpan.OperationName() == httpGRPCHandleMethod {
+				parentSpan.SetTag("http.request.method", r.Method)
+				parentSpan.SetTag("http.route", matchedRoute)
+				parentSpan.SetTag("url.path", r.RequestURI)
+			}
+		}
+
+		childSpanName := httpServerSpanName(r.Method, matchedRoute)
+		startSpanOpts := []opentracing.StartSpanOption{
+			ext.SpanKindRPCServer,
+			opentracing.Tag{Key: string(ext.Component), Value: "net/http"},
+			opentracing.Tag{Key: "http.request.method", Value: r.Method},
+			opentracing.Tag{Key: "http.route", Value: matchedRoute},
+			opentracing.Tag{Key: "url.path", Value: r.RequestURI},
+		}
+		childSpan, _ := opentracing.StartSpanFromContextWithTracer(ctx, tracer, childSpanName, startSpanOpts...)
+		defer childSpan.Finish()
```

Comment:
What happens if a HTTP request arrives directly (ie. not through httpgrpc)? 

Do we end up with a duplicate span? (I'd expect standard HTTP requests to already have a span created for them.)

Reply:
So it does create two HTTP spans, though I think this may actually be useful because we will get the same span child span regardless of whether the parent is an HTTP or gRPC request.

The parent HTTP span is not identical to the child HTTP span we're creating, because it's using default opentracing middleware, which adds some info like the status code and user agent (and opentracing has a different span naming convention).

Reply:
I think it would be confusing to see two spans for the non-httpgrpc case, so only emitting the span when it's a httpgrpc request makes sense to me.

>The parent HTTP span is not identical to the child HTTP span we're creating, because it's using default opentracing middleware, which adds some info like the status code and user agent (and opentracing has a different span naming convention).

Should we also attach similar info? I imagine the status code in particular would be very useful.